### PR TITLE
[Author] Add fpo w/ npm auto-update via single package.json

### PR DESCRIPTION
**Pull request by getify**
Pull request for issue: #10811
Related issue(s): # #

Checklist for **Pull request** or **lib adding request issue** follows the conventions.

Note that if you are using a distribution purpose repository/package, please also provide the url and other related info like popularity of the source code repo/package.

# Profile of the lib
 * Git repository (required): https://github.com/getify/fpo
 * Official website (optional, not the repository):
 * NPM package url (optional): https://www.npmjs.com/package/fpo
 * License and its reference: MIT https://github.com/getify/fpo/blob/master/LICENSE.txt
 * GitHub / Bitbucket popularity (required):
   - Count of watchers: 13
   - Count of stars: 248
   - Count of forks: 11
 * NPM download stats (optional):
   - Downloads in the last day: 1
   - Downloads in the last week: 2
   - Downloads in the last month: 198

# Essential checklist
 * [x] I'm the author of this library
   * [x] I would like to add link to the page of this library on CDNJS on website / readme
 * [x] This lib was not found on cdnjs repo
 * [x] No already exist / duplicated issue and PR
 * [x] The lib has notable popularity
   * [x] More than 200 Stars on GitHub
   * [ ] More than 800 downloads stats per month on npm registry
 * [x] Project has public repository on famous online hosting platform (or been hosted on npm)

# Auto-update checklist
 * [ ] Has valid tags for each versions (for git auto-update)
 * [ ] Auto-update setup
 * [ ] Auto-update target/source is valid.
 * [ ] Auto-update filemap is correct.

# Git commit checklist
 * [ ] The first line of commit message is less then 50 chars, be clean and clear, easy to understand.
 * [ ] The parent of the commit(s) in the PR is not old than 3 days.
 * [ ] Pull request is sending from a non-master branch with meaningful name.
 * [ ] Separate unrelated changes into different commits.
 * [ ] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.
 * [ ] Close corresponding issue in commit message
 * [ ] Mention related issue(s), people in commit message, comment.

### CODE REVIEW

**Comment by getify** [comment]
I did the process through the github UI, not on the command line. I followed the steps laid out here: https://github.com/cdnjs/cdnjs/issues/10811#issuecomment-292791860

Including the exact commit message as stated there.

Please let me know what the commit message is supposed to say, and how to fix this in the github UI?

**Comment by extend1994** [comment]
@getify The only difference between the commit we want and yours is the heading of the commit message. The heading would be `Add fpo w/ npm auto-update via single package.json` not the current one `Add fpo` and the body would be `close #1081`. Unfortunately, you can't directly modify commit message on GitHub GUI, because they don't support this. Do you mind I modify it for you?

**Comment by getify** [comment]
>  Do you mind I modify it for you?

No, I absolutely don't mind. Please change as you see fit. It wasn't clear to me that the two lines in the thread message were supposed to be "first line" = "description", "second line" = "body". Sorry for my confusion.

**Comment by getify** [comment]
And btw, when I look at the other commits in the master branch, they all have commit descriptions like "Add XYZ v3.4.5" (and not anything like "w/ npm auto-update via single package.json" in them), so that's where my "Add FPO" came from. :)

**Comment by extend1994** [comment]
Sorry for confusing you 😓, maybe you can take a look after I make the commits, which might make things more clearer!

The commits you see in master branch were added by the other method [adding a new library with its assets](https://github.com/cdnjs/cdnjs/blob/master/CONTRIBUTING.md#d-adding-a-new-library-with-its-assets). More disk spaces would be used if you use the method [adding a new library with its assets](https://github.com/cdnjs/cdnjs/blob/master/CONTRIBUTING.md#d-adding-a-new-library-with-its-assets). So we provide the method [adding a new library by a single package.json](https://github.com/cdnjs/cdnjs/blob/master/CONTRIBUTING.md#e-adding-a-new-library-by-a-single-packagejson) to let contributors be able to simply create a single package.json to contribute CDNJS. That's why the commit message is slightly different.

**Comment by extend1994** [comment]
@getify Because the file structure in v0.0.1 ~ v2.5.0 is different from the current one. We need to manually add them if they also can be used in a browser (I will do it for you). So the question is that can we use fpo v0.0.1 ~ v2.5.0 in a browser? If yes, which files should we host? 
I tried to figure out it on my own according to README.md, which said that users need to build the files with build-core.js. But I found fpo.js is the same as lib/fpo.src.js in v0.0.1 ~ v2.5.0 except v2.0.0. In v2.0.0, after building fpo.js with build-core.js, fpo.js would be the same as lib/fpo.src.js. So I want to confirm this with you.

**Comment by getify** [comment]
Is it possible to just have the CDN only serve from 6.0 upward of the package, and ignore everything before that? Truthfully, it was in a lot of flux earlier on and nobody was really using it in those versions anyway. It's much more stable now in terms of file structure, etc.

**Comment by PeterDaveHello** [comment]
@getify Of course we can do that!

**Comment by sufuf3** [comment]
Help modify the commit message.

@getify as your comment, would you update `.gitignore` to exclude the versions before v5.2.3(include v5.2.3)?
Just add the lines as the followings into `.gitignore`
```
/ajax/libs/fpo/0*
/ajax/libs/fpo/1*
/ajax/libs/fpo/2*
/ajax/libs/fpo/3*
/ajax/libs/fpo/4*
/ajax/libs/fpo/5*
```
And the example of the commit message as the followings (following git commit message: 50/72 formatting):
```
Update .gitignore for fpo unstable versions, cc #10811

Because it was in a lot of flux earlier on and nobody was really using
it in those versions anyway. It's much more stable now in terms of
file structure, etc.
```

Thank you.

**Comment by getify** [comment]
done :)

**Comment by PeterDaveHello** [comment]
@sufuf3 1. Can't these patterns be merged?
2. What would happen to version 10, 11, and every versions starting with 0~5 prefix?

**Comment by sufuf3** [comment]
Oh, sorry for not detail comment.
I mean [`cdnjs/.gitignore`](https://github.com/cdnjs/cdnjs/blob/master/.gitignore) this file not `cdnjs/ajax/libs/fpo/.gitignore`.

**Comment by sufuf3** [comment]
Sorry the pattern needs to change to 
```
/ajax/libs/fpo/0.*
/ajax/libs/fpo/1.*
/ajax/libs/fpo/2.*
/ajax/libs/fpo/3.*
/ajax/libs/fpo/4.*
/ajax/libs/fpo/5.*
```

**Comment by PeterDaveHello** [comment]
@sufuf3 Can't we merged these rules?

**Comment by sufuf3** [comment]
I think yes, just need to use one line.
```
/ajax/libs/fpo/[3-5].*
```
I learn from this line https://github.com/cdnjs/cdnjs/blob/master/.gitignore#L223

**Comment by PeterDaveHello** [comment]
@sufuf3 so where is your origin 0~2? Or we just don't need that?

**Comment by sufuf3** [comment]
Because the current pattern can't get the files from v0.* ~ v2.* 
So I exclude them.

Sorry for not mention this.

**Comment by sufuf3** [comment]
@getify Would you update `cdnjs/.gitignore` and remove `cdnjs/ajax/libs/fpo/.gitignore` in the second commit?
And add the line `/ajax/libs/fpo/[3-5].*` into `cdnjs/.gitignore`.
I sorry for not mention the detail about `.gitignore`.

If this is a little bit bother you, I can help you to do this.

Thank you for your contribution.

**Comment by sufuf3** [comment]
@getify Would you update the `cdnjs/.gitignore` file?
Or let me help you do this?
I think this PR almost finish except for this part.

Thank you.

**Comment by getify** [comment]
@sufuf3 I think I've made the requested changes. But now there's conflicts to resolve. Not sure how to proceed.

**Comment by getify** [comment]
@sufuf3 maybe I should just delete the whole branch and start over (or give up).

File: .gitignore [hunk 1]
```diff
@@ -223,6 +223,7 @@ node_modules/
 /ajax/libs/slim.js/2.[0-5].*/
 /ajax/libs/europa/4.0.0-alpha*/
 /ajax/libs/inferno-compat/*-alpha.*/
+/ajax/libs/fpo/[3-5].*
```
Comment (PeterDaveHello) [inline .gitignore]: @getify I think you can simply move this line to somewhere else like line 200 and it'll fix the conflict, don't need to give up :)

**Comment by sufuf3** [comment]
@getify You are almost there, you can move `/ajax/libs/fpo/[3-5].*` to somewhere.
If you need help, please let me know.
Let's finish this PR. 👍  
Thank you for your contribution.

**Pull request merged**